File: 2.Add-Two-Numbers.go
// https://leetcode.com/problems/add-two-numbers/
//
// algorithms
// Medium (30.1%)
// Total Accepted:    707,610
// Total Submissions: 2,350,674

package leetcode

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	arr1, arr2 := genIntArr(l1), genIntArr(l2)
	len1, len2 := len(arr1), len(arr2)
	res := &ListNode{}
	tmp := res
	carry := 0

	for i := 0; i < max(len1, len2); i++ {
		bit := carry
		if i < len1 {
			bit += arr1[len1-1-i]
		}
		if i < len2 {
			bit += arr2[len2-1-i]
		}
		carry = bit / 10
		tmp.Next = &ListNode{
			Val: bit % 10,
		}
		tmp = tmp.Next
	}

	if carry > 0 {
		tmp.Next = &ListNode{
			Val: carry,
		}
	}

	return res.Next
}

func genIntArr(l *ListNode) []int {
	var res []int

	for l != nil {
		res = append([]int{l.Val}, res...)
		l = l.Next
	}

	return res
}

func max(a, b int) int {
	if a < b {
		return b
	}
	return a
}

File: 2.Add-Two-Numbers.py
# https://leetcode.com/problems/add-two-numbers/
#
# algorithms
# Medium (31.07%)
# Total Accepted:    864,234
# Total Submissions: 2,781,507


# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        if not l1:
            return l2
        if not l2:
            return l1
        
        res = l1
        pre = None
        
        carry = 0
        while l1 and l2:
            l1.val = l1.val + l2.val + carry
            if l1.val >= 10:
                carry = 1
                l1.val -= 10
            else:
                carry = 0
            pre = l1
            l1 = l1.next
            l2 = l2.next
        
        if l2:
            pre.next = l2
            l1 = l2
        
        while l1:
            l1.val += carry
            if l1.val >= 10:
                carry = 1
                l1.val -= 10
            else:
                carry = 0
            pre = l1
            l1 = l1.next
        
        if carry == 1:
            pre.next = ListNode(1)
        
        return res
        